File: src/panel_order/detector.py
from __future__ import annotations

from dataclasses import dataclass

import cv2
import numpy as np

from .types import Panel, Rect


@dataclass(slots=True)
class PanelDetectionConfig:
    """Thresholds and search ranges used by the panel detector."""

    adaptive_block_size: int = 31
    adaptive_c: int = 10
    close_kernel_size: int = 5
    min_area_ratio: float = 0.01
    max_area_ratio: float = 0.92
    min_rectangularity: float = 0.82
    max_polygon_vertices: int = 10
    min_aspect_ratio: float = 0.12
    max_aspect_ratio: float = 8.0
    min_border_score: float = 0.35
    refine_margin_ratio: float = 0.18
    max_refine_margin_px: int = 80
    border_band_px: int = 5
    overlap_iou_threshold: float = 0.55
    contain_threshold: float = 0.9


class PanelDetector:
    """Detect rectangular manga panels from a single page image."""

    def __init__(self, config: PanelDetectionConfig | None = None) -> None:
        """Create a detector with either the provided config or default thresholds."""
        self.config = config or PanelDetectionConfig()

    def detect(self, image: np.ndarray) -> list[Panel]:
        """Run the full panel detection pipeline on an input image."""
        # Stage 1: Normalize the input into a grayscale image and extract panel lines.
        gray = self._to_grayscale(image)
        line_mask = self._build_line_mask(gray)

        # Stage 2: Convert contours into panel candidates with geometric filtering.
        raw_candidates = self._extract_candidates(gray, line_mask)

        # Stage 3: Remove duplicates and assign stable panel ids in page order.
        panels = self._deduplicate(raw_candidates)
        ordered = sorted(panels, key=lambda panel: (panel.rect.y, panel.rect.x))
        return [Panel(panel_id=index, rect=panel.rect, score=panel.score) for index, panel in enumerate(ordered)]

    def _to_grayscale(self, image: np.ndarray) -> np.ndarray:
        """Convert a color page image into grayscale when needed."""
        if image.ndim == 2:
            return image
        return cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

    def _build_line_mask(self, gray: np.ndarray) -> np.ndarray:
        """Build a binary mask that highlights panel borders and gutters."""
        # Blur first so adaptive thresholding reacts to panel lines instead of screentone noise.
        blurred = cv2.GaussianBlur(gray, (5, 5), 0.0)
        binary = cv2.adaptiveThreshold(
            blurred,
            255,
            cv2.ADAPTIVE_THRESH_GAUSSIAN_C,
            cv2.THRESH_BINARY_INV,
            self.config.adaptive_block_size,
            self.config.adaptive_c,
        )

        # Close small gaps so broken border strokes can still form stable contours.
        kernel = np.ones((self.config.close_kernel_size, self.config.close_kernel_size), dtype=np.uint8)
        return cv2.morphologyEx(binary, cv2.MORPH_CLOSE, kernel, iterations=1)

    def _extract_candidates(self, gray: np.ndarray, line_mask: np.ndarray) -> list[Panel]:
        """Extract panel candidates from the line mask using contour geometry."""
        contours, _hierarchy = cv2.findContours(line_mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
        page_area = gray.shape[0] * gray.shape[1]
        candidates: list[Panel] = []

        for contour in contours:
            x, y, width, height = cv2.boundingRect(contour)
            area = width * height

            # Filter out contours that are too small to be panels or too large to be useful.
            if area < page_area * self.config.min_area_ratio or area > page_area * self.config.max_area_ratio:
                continue

            aspect_ratio = float(width) / float(max(height, 1))
            if aspect_ratio < self.config.min_aspect_ratio or aspect_ratio > self.config.max_aspect_ratio:
                continue

            # Keep only contours that are close to rectangular panel outlines.
            contour_area = float(cv2.contourArea(contour))
            rectangularity = contour_area / float(max(area, 1))
            if rectangularity < self.config.min_rectangularity:
                continue

            perimeter = cv2.arcLength(contour, True)
            polygon = cv2.approxPolyDP(contour, 0.02 * perimeter, True)
            if len(polygon) > self.config.max_polygon_vertices:
                continue

            # Refine the raw contour bounds by snapping each side toward stronger border lines.
            refined = self._refine_rect(line_mask, Rect(x, y, width, height))
            border_score = self._border_score(line_mask, refined)
            if border_score < self.config.min_border_score:
                continue

            score = rectangularity * 0.6 + border_score * 0.4
            candidates.append(Panel(panel_id=-1, rect=refined, score=float(score)))

        return candidates

    def _refine_rect(self, line_mask: np.ndarray, rect: Rect) -> Rect:
        """Refine a rectangle by searching for stronger borders near each side."""
        margin_x = min(self.config.max_refine_margin_px, max(6, int(rect.width * self.config.refine_margin_ratio)))
        margin_y = min(self.config.max_refine_margin_px, max(6, int(rect.height * self.config.refine_margin_ratio)))
        band = self.config.border_band_px

        # Search inward from each edge so loose contour boxes collapse onto visible panel borders.
        left = self._best_vertical_boundary(line_mask, rect.x, rect.y, rect.height, +1, margin_x, band)
        right = self._best_vertical_boundary(line_mask, rect.right - band, rect.y, rect.height, -1, margin_x, band) + band
        top = self._best_horizontal_boundary(line_mask, rect.y, rect.x, rect.width, +1, margin_y, band)
        bottom = self._best_horizontal_boundary(line_mask, rect.bottom - band, rect.x, rect.width, -1, margin_y, band) + band

        if right <= left + 10 or bottom <= top + 10:
            return rect
        return Rect(left, top, right - left, bottom - top)

    def _best_vertical_boundary(
        self,
        line_mask: np.ndarray,
        edge_x: int,
        top: int,
        height: int,
        direction: int,
        max_shift: int,
        band: int,
    ) -> int:
        """Find the strongest nearby vertical border band for one rectangle side."""
        height_limit, width_limit = line_mask.shape
        top = max(0, top)
        bottom = min(height_limit, top + height)
        best_x = int(np.clip(edge_x, 0, width_limit - band))
        best_value = -1.0

        # Prefer strong border evidence, but penalize long shifts so boundaries stay near the seed box.
        for shift in range(max_shift + 1):
            x = int(np.clip(edge_x + direction * shift, 0, width_limit - band))
            score = float(line_mask[top:bottom, x : x + band].mean()) / 255.0
            score -= 0.3 * (float(shift) / float(max(max_shift, 1)))
            if score > best_value:
                best_value = score
                best_x = x

        return best_x

    def _best_horizontal_boundary(
        self,
        line_mask: np.ndarray,
        edge_y: int,
        left: int,
        width: int,
        direction: int,
        max_shift: int,
        band: int,
    ) -> int:
        """Find the strongest nearby horizontal border band for one rectangle side."""
        height_limit, width_limit = line_mask.shape
        left = max(0, left)
        right = min(width_limit, left + width)
        best_y = int(np.clip(edge_y, 0, height_limit - band))
        best_value = -1.0

        # Use the same border search strategy vertically for the top and bottom edges.
        for shift in range(max_shift + 1):
            y = int(np.clip(edge_y + direction * shift, 0, height_limit - band))
            score = float(line_mask[y : y + band, left:right].mean()) / 255.0
            score -= 0.3 * (float(shift) / float(max(max_shift, 1)))
            if score > best_value:
                best_value = score
                best_y = y

        return best_y

    def _border_score(self, line_mask: np.ndarray, rect: Rect) -> float:
        """Score how strongly all four rectangle sides align with panel borders."""
        band = min(self.config.border_band_px, rect.width, rect.height)
        if band <= 0:
            return 0.0
        top = float(line_mask[rect.y : rect.y + band, rect.x : rect.right].mean()) / 255.0
        bottom = float(line_mask[rect.bottom - band : rect.bottom, rect.x : rect.right].mean()) / 255.0
        left = float(line_mask[rect.y : rect.bottom, rect.x : rect.x + band].mean()) / 255.0
        right = float(line_mask[rect.y : rect.bottom, rect.right - band : rect.right].mean()) / 255.0
        return (top + bottom + left + right) / 4.0

    def _deduplicate(self, candidates: list[Panel]) -> list[Panel]:
        """Remove overlapping candidates while keeping the highest-quality box."""
        kept: list[Panel] = []
        ranked = sorted(candidates, key=lambda panel: (panel.score, panel.rect.area), reverse=True)
        for candidate in ranked:
            should_keep = True
            for existing in kept:
                # Suppress candidates that are almost contained by, or heavily overlap with, a kept box.
                if existing.rect.contains_ratio(candidate.rect) >= self.config.contain_threshold:
                    should_keep = False
                    break
                if candidate.rect.iou(existing.rect) >= self.config.overlap_iou_threshold:
                    should_keep = False
                    break
            if should_keep:
                kept.append(candidate)

        return sorted(kept, key=lambda panel: (panel.rect.y, panel.rect.x))

File: web/src/core/panel_order/detector.ts
import { buildPanelLayout } from "./layout";
import { loadOpenCv } from "./opencv";
import type { PanelAnalysis, PanelBox, PanelRect } from "./types";

type PanelDetectionConfig = {
  adaptiveBlockSize: number;
  adaptiveC: number;
  closeKernelSize: number;
  minAreaRatio: number;
  maxAreaRatio: number;
  minRectangularity: number;
  maxPolygonVertices: number;
  minAspectRatio: number;
  maxAspectRatio: number;
  minBorderScore: number;
  refineMarginRatio: number;
  maxRefineMarginPx: number;
  borderBandPx: number;
  overlapIouThreshold: number;
  containThreshold: number;
};

const DEFAULT_CONFIG: PanelDetectionConfig = {
  adaptiveBlockSize: 31,
  adaptiveC: 10,
  closeKernelSize: 5,
  minAreaRatio: 0.01,
  maxAreaRatio: 0.92,
  minRectangularity: 0.82,
  maxPolygonVertices: 10,
  minAspectRatio: 0.12,
  maxAspectRatio: 8.0,
  minBorderScore: 0.35,
  refineMarginRatio: 0.18,
  maxRefineMarginPx: 80,
  borderBandPx: 5,
  overlapIouThreshold: 0.55,
  containThreshold: 0.9,
};

function rectRight(rect: PanelRect): number {
  return rect.x + rect.width;
}

function rectBottom(rect: PanelRect): number {
  return rect.y + rect.height;
}

function rectArea(rect: PanelRect): number {
  return rect.width * rect.height;
}

function rectIntersection(left: PanelRect, right: PanelRect): PanelRect | null {
  const x = Math.max(left.x, right.x);
  const y = Math.max(left.y, right.y);
  const r = Math.min(rectRight(left), rectRight(right));
  const b = Math.min(rectBottom(left), rectBottom(right));
  if (r <= x || b <= y) {
    return null;
  }
  return { x, y, width: r - x, height: b - y };
}

function rectIoU(left: PanelRect, right: PanelRect): number {
  const overlap = rectIntersection(left, right);
  if (!overlap) {
    return 0;
  }
  const unionArea = rectArea(left) + rectArea(right) - rectArea(overlap);
  return rectArea(overlap) / Math.max(unionArea, 1);
}

function containsRatio(container: PanelRect, candidate: PanelRect): number {
  const overlap = rectIntersection(container, candidate);
  if (!overlap) {
    return 0;
  }
  return rectArea(overlap) / Math.max(rectArea(candidate), 1);
}

function imageBitmapToImageData(bitmap: ImageBitmap): ImageData {
  const canvas = document.createElement("canvas");
  canvas.width = bitmap.width;
  canvas.height = bitmap.height;
  const context = canvas.getContext("2d", { willReadFrequently: true });
  if (!context) {
    throw new Error("Failed to create a 2D canvas for panel detection.");
  }
  context.drawImage(bitmap, 0, 0);
  return context.getImageData(0, 0, bitmap.width, bitmap.height);
}

function buildLineMask(cv: any, gray: any, config: PanelDetectionConfig): any {
  const blurred = new cv.Mat();
  const binary = new cv.Mat();
  const lineMask = new cv.Mat();
  const kernel = cv.Mat.ones(config.closeKernelSize, config.closeKernelSize, cv.CV_8U);

  try {
    // Blur first so adaptive thresholding reacts to borders instead of screentone noise.
    cv.GaussianBlur(gray, blurred, new cv.Size(5, 5), 0, 0, cv.BORDER_DEFAULT);
    cv.adaptiveThreshold(
      blurred,
      binary,
      255,
      cv.ADAPTIVE_THRESH_GAUSSIAN_C,
      cv.THRESH_BINARY_INV,
      config.adaptiveBlockSize,
      config.adaptiveC,
    );
    // Close small gaps so broken borders still form usable contours.
    cv.morphologyEx(binary, lineMask, cv.MORPH_CLOSE, kernel, new cv.Point(-1, -1), 1);
    return lineMask.clone();
  } finally {
    blurred.delete();
    binary.delete();
    lineMask.delete();
    kernel.delete();
  }
}

function sliceMean(cv: any, mat: any, x: number, y: number, width: number, height: number): number {
  const roi = mat.roi(new cv.Rect(x, y, width, height));
  try {
    const mean = cv.mean(roi);
    return mean[0] / 255.0;
  } finally {
    roi.delete();
  }
}

function bestVerticalBoundary(
  cv: any,
  lineMask: any,
  edgeX: number,
  top: number,
  height: number,
  direction: number,
  maxShift: number,
  band: number,
): number {
  const heightLimit = lineMask.rows as number;
  const widthLimit = lineMask.cols as number;
  const clampedTop = Math.max(0, top);
  const bottom = Math.min(heightLimit, clampedTop + height);
  let bestX = Math.max(0, Math.min(edgeX, widthLimit - band));
  let bestScore = -Infinity;

  for (let shift = 0; shift <= maxShift; shift += 1) {
    const x = Math.max(0, Math.min(edgeX + direction * shift, widthLimit - band));
    const score = sliceMean(cv, lineMask, x, clampedTop, band, bottom - clampedTop) - 0.3 * (shift / Math.max(maxShift, 1));
    if (score > bestScore) {
      bestScore = score;
      bestX = x;
    }
  }
  return bestX;
}

function bestHorizontalBoundary(
  cv: any,
  lineMask: any,
  edgeY: number,
  left: number,
  width: number,
  direction: number,
  maxShift: number,
  band: number,
): number {
  const heightLimit = lineMask.rows as number;
  const widthLimit = lineMask.cols as number;
  const clampedLeft = Math.max(0, left);
  const right = Math.min(widthLimit, clampedLeft + width);
  let bestY = Math.max(0, Math.min(edgeY, heightLimit - band));
  let bestScore = -Infinity;

  for (let shift = 0; shift <= maxShift; shift += 1) {
    const y = Math.max(0, Math.min(edgeY + direction * shift, heightLimit - band));
    const score = sliceMean(cv, lineMask, clampedLeft, y, right - clampedLeft, band) - 0.3 * (shift / Math.max(maxShift, 1));
    if (score > bestScore) {
      bestScore = score;
      bestY = y;
    }
  }
  return bestY;
}

function refineRect(cv: any, lineMask: any, rect: PanelRect, config: PanelDetectionConfig): PanelRect {
  const marginX = Math.min(config.maxRefineMarginPx, Math.max(6, Math.round(rect.width * config.refineMarginRatio)));
  const marginY = Math.min(config.maxRefineMarginPx, Math.max(6, Math.round(rect.height * config.refineMarginRatio)));
  const band = config.borderBandPx;

  const left = bestVerticalBoundary(cv, lineMask, rect.x, rect.y, rect.height, +1, marginX, band);
  const right = bestVerticalBoundary(cv, lineMask, rectRight(rect) - band, rect.y, rect.height, -1, marginX, band) + band;
  const top = bestHorizontalBoundary(cv, lineMask, rect.y, rect.x, rect.width, +1, marginY, band);
  const bottom = bestHorizontalBoundary(cv, lineMask, rectBottom(rect) - band, rect.x, rect.width, -1, marginY, band) + band;

  if (right <= left + 10 || bottom <= top + 10) {
    return rect;
  }
  return {
    x: left,
    y: top,
    width: right - left,
    height: bottom - top,
  };
}

function borderScore(cv: any, lineMask: any, rect: PanelRect, config: PanelDetectionConfig): number {
  const band = Math.min(config.borderBandPx, rect.width, rect.height);
  if (band <= 0) {
    return 0;
  }
  const top = sliceMean(cv, lineMask, rect.x, rect.y, rect.width, band);
  const bottom = sliceMean(cv, lineMask, rect.x, rectBottom(rect) - band, rect.width, band);
  const left = sliceMean(cv, lineMask, rect.x, rect.y, band, rect.height);
  const right = sliceMean(cv, lineMask, rectRight(rect) - band, rect.y, band, rect.height);
  return (top + bottom + left + right) / 4.0;
}

function deduplicateCandidates(candidates: PanelBox[], config: PanelDetectionConfig): PanelBox[] {
  const kept: PanelBox[] = [];
  const ranked = [...candidates].sort((left, right) => {
    if (left.score !== right.score) {
      return right.score - left.score;
    }
    return rectArea(right.rect) - rectArea(left.rect);
  });

  for (const candidate of ranked) {
    let shouldKeep = true;
    for (const existing of kept) {
      if (containsRatio(existing.rect, candidate.rect) >= config.containThreshold) {
        shouldKeep = false;
        break;
      }
      if (rectIoU(candidate.rect, existing.rect) >= config.overlapIouThreshold) {
        shouldKeep = false;
        break;
      }
    }
    if (shouldKeep) {
      kept.push(candidate);
    }
  }

  return kept.sort((left, right) => {
    if (left.rect.y !== right.rect.y) {
      return left.rect.y - right.rect.y;
    }
    return left.rect.x - right.rect.x;
  });
}

export async function analyzePanels(bitmap: ImageBitmap): Promise<PanelAnalysis> {
  const cv = (await loadOpenCv()) as any;
  const imageData = imageBitmapToImageData(bitmap);
  const source = cv.matFromImageData(imageData);
  const gray = new cv.Mat();
  const hierarchy = new cv.Mat();
  const contours = new cv.MatVector();
  const config = DEFAULT_CONFIG;

  try {
    cv.cvtColor(source, gray, cv.COLOR_RGBA2GRAY);
    const lineMask = buildLineMask(cv, gray, config);
    try {
      // Find contour candidates before filtering them with panel-specific geometry tests.
      cv.findContours(lineMask, contours, hierarchy, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE);
      const pageArea = gray.rows * gray.cols;
      const rawCandidates: PanelBox[] = [];

      for (let index = 0; index < contours.size(); index += 1) {
        const contour = contours.get(index);
        const polygon = new cv.Mat();
        try {
          const bounds = cv.boundingRect(contour);
          const rect: PanelRect = {
            x: bounds.x,
            y: bounds.y,
            width: bounds.width,
            height: bounds.height,
          };
          const area = rectArea(rect);
          if (area < pageArea * config.minAreaRatio || area > pageArea * config.maxAreaRatio) {
            continue;
          }

          const aspectRatio = rect.width / Math.max(rect.height, 1);
          if (aspectRatio < config.minAspectRatio || aspectRatio > config.maxAspectRatio) {
            continue;
          }

          const contourArea = cv.contourArea(contour);
          const rectangularity = contourArea / Math.max(area, 1);
          if (rectangularity < config.minRectangularity) {
            continue;
          }

          const perimeter = cv.arcLength(contour, true);
          cv.approxPolyDP(contour, polygon, 0.02 * perimeter, true);
          if (polygon.rows > config.maxPolygonVertices) {
            continue;
          }

          const refinedRect = refineRect(cv, lineMask, rect, config);
          const refinedBorderScore = borderScore(cv, lineMask, refinedRect, config);
          if (refinedBorderScore < config.minBorderScore) {
            continue;
          }

          rawCandidates.push({
            panelId: -1,
            rect: refinedRect,
            score: rectangularity * 0.6 + refinedBorderScore * 0.4,
          });
        } finally {
          polygon.delete();
          contour.delete();
        }
      }

      const panels = deduplicateCandidates(rawCandidates, config).map((panel, index) => ({
        ...panel,
        panelId: index,
      }));
      if (panels.length === 0) {
        throw new Error("No manga panels were detected.");
      }

      const { layoutTree, readingOrder } = buildPanelLayout(panels);
      return {
        imageWidth: bitmap.width,
        imageHeight: bitmap.height,
        panels,
        layoutTree,
        readingOrder,
      };
    } finally {
      lineMask.delete();
    }
  } finally {
    source.delete();
    gray.delete();
    hierarchy.delete();
    contours.delete();
  }
}
